### PR TITLE
CLN: BlockManager.get_slice require only slice arg

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

non-slice objects are currently getting past mypy.  This ensures we only get slices, will open up some optimization options.

Also did some docstring/typing cleanup in the neighborhood.

<b>update</b>: updated to implement Index._getitem_slice.  Cuts about 10% off of the benchmark discussed in https://github.com/pandas-dev/pandas/pull/40171#issuecomment-790219422 (as compared to current master, not then-master)

### CODE REVIEW

File: pandas/core/internals/managers.py [hunk 1]
```diff
@@ -1604,14 +1609,23 @@ def _blklocs(self):
         """ compat with BlockManager """
         return None
 
+    def getitem_mgr(self, indexer) -> SingleBlockManager:
```
Comment (jreback) [inline pandas/core/internals/managers.py:1612]: shouldn't this exist for BlockManager and just  be no-op?

**Review by jreback — CHANGES_REQUESTED** [review]
this is quite a lot of added code paths in index, are these really all tested?

File: pandas/core/internals/managers.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/internals/managers.py:1612]: its only ever used by Series

**Comment by jbrockmendel** [comment]
> this is quite a lot of added code paths in index, are these really all tested?

yes

**Review by jreback — APPROVED** [review]

**Comment by jbrockmendel** [comment]
> we have sufficient asv's to cover? (assume we do)

yes.  this is motivated by the one discussed in joris's libreduction PR

**Comment by jreback** [comment]
lgtm. can you rebase & ping on green.

**Comment by jbrockmendel** [comment]
Ping

**Pull request merged**